### PR TITLE
implement `gt rename-branch`

### PR DESCRIPTION
**Pull request by charlierudolph**

### CODE REVIEW

File: lib/git/config.go [hunk 1]
```diff
@@ -35,10 +42,12 @@ func CompileAncestorBranches(branchName string) (result []string) {
...
-	configs := util.GetCommandOutput("git", "config", "--get-regexp", "^git-town-branch\\..*\\.ancestors$")
-	for _, config := range strings.Split(configs, "\n") {
-		splitConfig := strings.Split(config, " ")
-		removeConfigurationValue(splitConfig[0])
+	configRegexp, err := regexp.Compile("^git-town-branch\\..*\\.ancestors$")
+	if err != nil {
+		log.Fatal("Error compiling ancestor branches configuration regular expression: ", err)
+	}
+	for _, key := range getConfigurationKeysMatching(configRegexp) {
+		removeConfigurationValue(key)
```
Comment (charlierudolph) [inline lib/git/config.go]: Had to change how this works as `git config --get-regexp "^git-town-branch\\..*\\.ancestors$"` returns an error code if there are it matches no keys. This was surfaced in one of the feature tests for rename-branch.

Same applies to a similar code block below

File: lib/git/config.go [hunk 2]
```diff
@@ -15,6 +15,13 @@ import (
 	"github.com/Originate/git-town/lib/util"
 )
 
+// AddPerennialBranch adds the given branch as a perennial branch
+func AddPerennialBranch(branchName string) {
+	oldValue := GetPerennialBranches()
+	newValue := append(oldValue, branchName)
```
Comment (kevgo) [inline lib/git/config.go]: There are too many temp variables here, which is a code smell. Also, `oldValue` doesn't describe much. I would inline `oldValue` and possibly also `newValue`. 

```go
setPerennialBranches(append(GetPerennialBranches(), branchName))

File: lib/git/config.go [hunk 3]
```diff
@@ -35,10 +42,12 @@ func CompileAncestorBranches(branchName string) (result []string) {
...
 func DeleteAllAncestorBranches() {
-	configs := util.GetCommandOutput("git", "config", "--get-regexp", "^git-town-branch\\..*\\.ancestors$")
-	for _, config := range strings.Split(configs, "\n") {
-		splitConfig := strings.Split(config, " ")
-		removeConfigurationValue(splitConfig[0])
+	configRegexp, err := regexp.Compile("^git-town-branch\\..*\\.ancestors$")
+	if err != nil {
+		log.Fatal("Error compiling ancestor branches configuration regular expression: ", err)
+	}
+	for _, key := range getConfigurationKeysMatching(configRegexp) {
```
Comment (kevgo) [inline lib/git/config.go]: What about making `getConfigurationKeysMatching` take in a string and compile the regex in there? The code to compile regex is duplicated in a number of places.

File: lib/git/config.go [hunk 4]
```diff
@@ -185,6 +195,13 @@ func IsPerennialBranch(branchName string) bool {
 	return util.DoesStringArrayContain(perennialBranches, branchName)
 }
 
+// RemovePerennialBranch removes the given branch as a perennial branch
+func RemovePerennialBranch(branchName string) {
+	oldValue := GetPerennialBranches()
+	newValue := util.RemoveStringElement(oldValue, branchName)
+	setPerennialBranches(newValue)
```
Comment (kevgo) [inline lib/git/config.go]: Inline all temp variables here

File: lib/util/util.go [hunk 5]
```diff
@@ -35,6 +35,17 @@ func DoesStringArrayContain(list []string, value string) bool {
 	return false
 }
 
+// RemoveStringElement returns a new string slice which is the given string slice
+// with the given string removed
+func RemoveStringElement(list []string, value string) (result []string) {
```
Comment (kevgo) [inline lib/util/util.go]: What about naming this `RemoveStringFromSlice`?

File: cmd/rename_branch.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+func checkRenameBranchPreconditions(args []string) (result renameBranchConfig) {
+	if len(args) == 1 {
+		result.OldBranchName = git.GetCurrentBranchName()
+		result.NewBranchName = args[0]
+	} else {
+		result.OldBranchName = args[0]
+		result.NewBranchName = args[1]
+	}
+	if git.IsMainBranch(result.OldBranchName) {
```
Comment (charlierudolph) [inline cmd/rename_branch.go]: done

File: cmd/rename_branch.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+		util.ExitWithErrorMessage("Cannot rename branch to current name.")
+	}
+	script.Fetch()
+	git.EnsureHasBranch(result.OldBranchName)
+	ensureBranchInSync(result.OldBranchName)
+	git.EnsureDoesNotHaveBranch(result.NewBranchName)
+	return
+}
+
+func ensureBranchInSync(branchName string) {
```
Comment (charlierudolph) [inline cmd/rename_branch.go]: done

File: features/git-town-rename-branch/perennial_branch/with_remote_origin/with_tracking_branch/with_unpulled_changes.feature [hunk 8]
```diff
@@ -14,12 +14,14 @@ Feature: git town-rename-branch: errors if renaming a perennial branch that has
...
+    When I run `gt rename-branch --force production renamed-production`
 
 
   Scenario: result
-    Then I get the error "The branch is not in sync with its tracking branch."
-    And I get the error "Run 'git town-sync production' to sync the branch."
+    Then it runs the commands
+      | BRANCH     | COMMAND           |
+      | production | git fetch --prune |
+    And I get the error "'production' is not in sync with its tracking branch. Sync the branches before renaming."
```
Comment (charlierudolph) [inline features/git-town-rename-branch/perennial_branch/with_remote_origin/with_tracking_branch/with_unpulled_changes.feature]: done

File: lib/util/util.go (hunk 5 above)
Comment (charlierudolph) [inline lib/util/util.go]: done

File: lib/git/config.go (hunk 4 above)
Comment (charlierudolph) [inline lib/git/config.go]: done

File: lib/git/config.go (hunk 3 above)
Comment (charlierudolph) [inline lib/git/config.go]: done

File: lib/git/config.go (hunk 2 above)
Comment (charlierudolph) [inline lib/git/config.go]: done

File: lib/git/config.go [hunk 9]
```diff
@@ -15,6 +15,13 @@ import (
 	"github.com/Originate/git-town/lib/util"
 )
 
+// AddPerennialBranch adds the given branch as a perennial branch
+func AddPerennialBranch(branchName string) {
```
Comment (charlierudolph) [inline lib/git/config.go]: We currently have the steps named in the same way. Would `AddAsPerennialBranch` make it more clear that we are not creating a new branch? I like how add/remove are opposite and would prefer to keep this and the related remove method in some form of sync. 

Our user interface uses `--add` and `--remove` and our current bash code uses `add_perennial_branch`.

File: features/git-town-rename-branch/current_branch.feature [hunk 10]
```diff
@@ -29,24 +37,32 @@ Feature: git town-rename-branch: rename current branch implicitly
...
 
 
   Scenario: rename perennial branch
     Given I am on the "production" branch
-    When I run `git town-rename-branch renamed-production`
+    When I run `gt rename-branch renamed-production`
     Then it runs no commands
-    And I get the error "The branch 'production' is not a feature branch."
-    And I get the error "Run 'git town-rename-branch production renamed-production -f' to force the rename, then reconfigure git-town on any other clones of this repo."
+    And I get the error "'production' is a perennial branch. Perennial branches require the '--force' option to be renamed."
```
Comment (charlierudolph) [inline features/git-town-rename-branch/current_branch.feature]: I'd favor a short explanation. What about

> 'production' is a perennial branch. Renaming a perennial branch typically requires other updates. If you are sure you want to do this, use '--force'.

**Comment by charlierudolph** [comment]
@kevgo all comments addressed

File: features/git-town-rename-branch/current_branch.feature (hunk 10 above)
Comment (kevgo) [inline features/git-town-rename-branch/current_branch.feature]: Shorter is always better! :)

File: lib/git/config.go (hunk 9 above)
Comment (kevgo) [inline lib/git/config.go]: Hmm, what about `AddToPerennialBranches` and `RemoveFromPerennialBranches`?

Comment (charlierudolph) [inline lib/git/config.go]: Sounds good

File: features/git-town-rename-branch/current_branch.feature (hunk 10 above)
Comment (charlierudolph) [inline features/git-town-rename-branch/current_branch.feature]: done

File: lib/git/config.go (hunk 9 above)
Comment (charlierudolph) [inline lib/git/config.go]: done

**Pull request merged**